platform/include/platform/CircularBuffer.h

```diff
@@ -98,63 +98,204 @@ class CircularBuffer {
+    void push(const mbed::Span<const T>& src)
+    {
+        push(src.data(), src.size());
+    }
+
+    /** Pop from the buffer.
+     *
+     * @param data Container to store the data to be popped from the buffer.
+     * @return True if data popped.
+     */
+    bool pop(T& data)
     {
         bool data_popped = false;
+
         core_util_critical_section_enter();
-        if (!empty()) {
-            data = _pool[_tail++];
-            if (_tail == BufferSize) {
-                _tail = 0;
+
+        if (!non_critical_empty()) {
+            data_popped = true;
+
+            data = _buffer[_tail];
+            _tail = incrementCounter(_tail);
+            _full = false;
+        }
+
+        core_util_critical_section_exit();
+
+        return data_popped;
+    }
+
+    /**
+     * Pop multiple elements from the buffer.
+     *
+     * @param dest The array which will receive the elements.
+     * @param len The number of elements to pop.
+     *
+     * @return The number of elements popped.
+     */
+    CounterType pop(T* dest, CounterType len)
+    {
+        CounterType data_popped = 0;
+
+        core_util_critical_section_enter();
+
+        if(!non_critical_empty()) {
+            /* make sure we only try to read as much as we have items present */
+            if (len > non_critical_size()) {
+                len = non_critical_size();
             }
+            data_popped = len;
+
+            /* items may be split by overlap, take only the number we have to the right of tail */
+            if ((_tail + data_popped) > BufferSize) {
+                data_popped = BufferSize - _tail;
+            }
+
+            std::copy(_buffer + _tail, _buffer + _tail + data_popped, dest);
+
+            /* if we looped over the end we may need to pop again */
+            CounterType left_to_pop = len - data_popped;
+
+            if (left_to_pop) {
+                std::copy(_buffer, _buffer + left_to_pop, dest + data_popped);
+
+                data_popped += left_to_pop;
+            }
+
+            _tail += data_popped;
+            _tail %= BufferSize;
             _full = false;
-            data_popped = true;
         }
+
         core_util_critical_section_exit();
+
         return data_popped;
     }
 
-    /** Check if the buffer is empty
+    /**
+     * Pop multiple elements from the buffer.
+     *
+     * @param dest The array which will receive the elements.
+     *
+     * @return The number of elements popped.
+     */
+    template<CounterType N>
+    CounterType pop(T (&dest)[N])
+    {
+        return pop(dest, N);
+    }
+
+    /**
+     * Pop multiple elements from the buffer.
+     *
+     * @param dest The span that contains the buffer that will be used to store the elements.
      *
-     * @return True if the buffer is empty, false if not
+     * @return The span with the size set to number of elements popped using the buffer passed in as the parameter.
+     */
+    mbed::Span<T> pop(mbed::Span<T>& dest)
```

Comment:
There's usually no reason to pass spans by reference rather than value - I've seen doing so mess with code generation in practice. If you did want to pass by reference, this should be `const span &` - the lack of const would be getting in the way of the automatic conversion from array to span, I think.

Return type seems reasonable to me, but ARM ABI does cause crummy code generation sometimes because of it, even with inlining. So returning just `popped` might be more efficient, but a span return is probably more idiomatic. No strong feelings.

Reply:
It's already const, I don't mind what is returned by pop either so unless either of you has opinions I'm leaving it.

Reply:
I see, it wasn't const before the span commit, must've been a change that didn't get saved

Reply:
Thanks for the feedback @kjbracey-arm .

If a span is returned then I believe it can be a `const span &` in input. The container in input is not altered and we return the content from the function as a result. 

For compiler optimizations,  I suppose one can use the _non span_ function if it matters.  

I initially asked Paul to return a `span` to allow writing code like this:

```c++
uint8_t chunk[50];

auto popped_chunk = queue.pop(chunk);
for (auto element : popped_chunk) {
    // do something with the element 
}

// or directly 
for (auto element : queue.pop(chunk)) {
    // do something with the element 
}
```

Reply:
OK, I'm passing it by value, thanks

Reply:
Certainly want to pass spans in by value normally. Doing it by reference isn't quite as hard to read as `int * const &ptr`, but it's just as weird, and does generate worse code.

I'm fine with span return.